src/google/protobuf/compiler/csharp/csharp_options.h

```diff
@@ -41,14 +41,29 @@ namespace csharp {
 
 // Generator options (used by csharp_generator.cc):
 struct Options {
-  Options() : file_extension(".cs"), base_namespace(""), generate_directories(false) {
+  Options() :
+      file_extension(".cs"),
+      base_namespace(""),
+      base_namespace_specified(false) {
   }
   // Extension of the generated file. Defaults to ".cs"
   string file_extension;
-  // Base namespace to use to create directory hierarchy. Defaults to ""
+  // Base namespace to use to create directory hierarchy. Defaults to "".
+  // This option allows the simple creation of a conventional C# file layout,
+  // where directories are created relative to a project-specific base
+  // namespace. For example, in a project with a base namespace of PetShop, a
+  // proto of user.proto with a C# namespace of PetShop.Model.Shared would
+  // generate Model/Shared/User.cs underneath the specified --csharp_out
+  // directory.
+  //
+  // If no base namespace is specified, all files are generated in the
+  // --csharp_out directory, with no subdirectories created automatically.
```

Comment:
> In the internal fix, I assume? I've just tried this locally and it just overwrites.

Sorry, I wasn't clear. I had switched to talking about C#, but I failed to mention that I had done that.

The `protoc` common code always verifies that a single invocation only produces each output file one time. If you violate that it will fail, not in the language-specific codegen but in the core compiler code.

The JavaScript codegen intentionally works around this behavior in order to emulate the behavior of the old Java-based codegen used inside Google. Inside the code generator it decides which set of output files to actually generate and then just skips the ones that would have gotten overwritten. So it implements the overwriting behavior without actually triggering protoc's overwrite-preventing code.

But since the C# codegen doesn't do this (I hope!) anyone using the C# codegen would get an error if they had two input files that would otherwise map to the same output file within a single invocation.

> Happy to merge for now?

Yep, sounds good.